File: ga/glm4_result_2024_6/result/tree_l_glm/24_tree_l.c
use std::io::{self, BufRead};
use std::collections::binary_search_tree::{Node, Tree};

fn main() {
    let mut tree = Tree::new();

    let stdin = io::stdin();
    let line = stdin.lock().lines().next().expect("Failed to read line").expect("Failed to read line");
    let tokens: Vec<&str> = line.split_whitespace().collect();
    for token in tokens {
        if let Ok(num) = token.parse::<i32>() {
            tree.insert(num);
        }
    }

    println!("Tree after insertion:");
    tree.print_in_order();

    tree.delete_even_numbers();

    println!("Tree after deleting even numbers:");
    tree.print_in_order();

    if !tree.is_empty() {
        println!("Min is {}, Max is {}", tree.find_min().unwrap(), tree.find_max().unwrap());
    } else {
        println!("Tree is empty");
    }
}

#[derive(Debug)]
struct Node {
    value: i32,
    left: Option<Box<Node>>,
    right: Option<Box<Node>>,
}

impl Node {
    fn new(value: i32) -> Node {
        Node {
            value,
            left: None,
            right: None,
        }
    }
}

type Tree = Option<Box<Node>>;

impl Tree {
    fn new() -> Tree {
        None
    }

    fn insert(&mut self, value: i32) {
        match self {
            Some(node) => {
                if value < node.value {
                    node.left.insert(value);
                } else if value > node.value {
                    node.right.insert(value);
                }
            }
            None => {
                *self = Some(Box::new(Node::new(value)));
            }
        }
    }

    fn delete_even_numbers(&mut self) {
        if let Some(node) = self {
            node.left = node.left.take().map(|left| {
                left.delete_even_numbers();
                left
            });
            node.right = node.right.take().map(|right| {
                right.delete_even_numbers();
                right
            });

            if node.value % 2 == 0 {
                if node.left.is_none() {
                    *self = node.right.take();
                } else if node.right.is_none() {
                    *self = node.left.take();
                } else {
                    let mut right = node.right.take().unwrap();
                    *self = node.left.take();
                    self.insert(right.value);
                }
            }
        }
    }

    fn find_min(&self) -> Option<i32> {
        match self {
            Some(node) => {
                if node.left.is_none() {
                    Some(node.value)
                } else {
                    node.left.find_min()
                }
            }
            None => None,
        }
    }

    fn find_max(&self) -> Option<i32> {
        match self {
            Some(node) => {
                if node.right.is_none() {
                    Some(node.value)
                } else {
                    node.right.find_max()
                }
            }
            None => None,
        }
    }

    fn print_in_order(&self) {
        match self {
            Some(node) => {
                node.left.print_in_order();
                println!("{}", node.value);
                node.right.print_in_order();
            }
            None => {}
        }
    }

    fn is_empty(&self) -> bool {
        self.is_none()
    }
}

File: ga/gpt3_5_result_2024_6/result/tree_l_gpt/24_tree_l.rs
use std::io::{self, BufRead};
use std::cmp::Ordering;

#[derive(Debug)]
struct TreeNode {
    element: i32,
    left: Option<Box<TreeNode>>,
    right: Option<Box<TreeNode>>,
}

impl TreeNode {
    fn new(element: i32) -> Self {
        TreeNode {
            element,
            left: None,
            right: None,
        }
    }
}

type SearchTree = Option<Box<TreeNode>>;

fn make_empty(tree: SearchTree) -> SearchTree {
    if let Some(mut node) = tree {
        node.left = make_empty(node.left);
        node.right = make_empty(node.right);
    }
    None
}

fn find(x: i32, tree: &SearchTree) -> Option<&TreeNode> {
    match tree {
        Some(node) => match x.cmp(&node.element) {
            Ordering::Less => find(x, &node.left),
            Ordering::Greater => find(x, &node.right),
            Ordering::Equal => Some(node),
        },
        None => None,
    }
}

fn find_min(tree: &SearchTree) -> Option<&TreeNode> {
    let mut current = tree;
    while let Some(node) = current {
        if let Some(left) = &node.left {
            current = Some(left);
        } else {
            return Some(node);
        }
    }
    None
}

fn find_max(tree: &SearchTree) -> Option<&TreeNode> {
    let mut current = tree;
    while let Some(node) = current {
        if let Some(right) = &node.right {
            current = Some(right);
        } else {
            return Some(node);
        }
    }
    None
}

fn insert(x: i32, tree: SearchTree) -> SearchTree {
    match tree {
        Some(mut node) => {
            match x.cmp(&node.element) {
                Ordering::Less => node.left = insert(x, node.left),
                Ordering::Greater => node.right = insert(x, node.right),
                Ordering::Equal => (), // Do nothing for duplicate elements
            }
            Some(node)
        },
        None => Some(Box::new(TreeNode::new(x))),
    }
}

fn delete(x: i32, tree: SearchTree) -> SearchTree {
    match tree {
        Some(mut node) => {
            match x.cmp(&node.element) {
                Ordering::Less => node.left = delete(x, node.left),
                Ordering::Greater => node.right = delete(x, node.right),
                Ordering::Equal => {
                    if node.left.is_none() {
                        return node.right;
                    } else if node.right.is_none() {
                        return node.left;
                    }
                    let min_right = find_min(&node.right);
                    if let Some(min_node) = min_right {
                        node.element = min_node.element;
                        node.right = delete(node.element, node.right);
                    }
                }
            }
            Some(node)
        },
        None => None,
    }
}

fn retrieve(node: &TreeNode) -> i32 {
    node.element
}

fn print_in_order(tree: &SearchTree) {
    match tree {
        Some(node) => {
            print_in_order(&node.left);
            println!("{}", node.element);
            print_in_order(&node.right);
        },
        None => (),
    }
}

fn delete_even_numbers(mut tree: SearchTree) -> SearchTree {
    if let Some(mut node) = tree {
        node.left = delete_even_numbers(node.left);
        node.right = delete_even_numbers(node.right);

        if node.element % 2 == 0 {
            tree = delete(node.element, tree);
        }
    }
    tree
}

fn main() {
    let mut tree: SearchTree = None;
    let stdin = io::stdin();
    let input = stdin.lock().lines().next().unwrap().unwrap();
    
    let mut has_input = false;
    for token in input.split_whitespace() {
        if let Ok(num) = token.parse::<i32>() {
            tree = insert(num, tree);
            has_input = true;
        }
    }

    if !has_input || tree.is_none() {
        println!("Tree is empty");
        return;
    }

    println!("Tree after insertion:");
    print_in_order(&tree);

    // Delete all even numbers
    tree = delete_even_numbers(tree);

    println!("Tree after deleting even numbers:");
    print_in_order(&tree);

    if let Some(tree_ref) = &tree {
        println!("Min is {}, Max is {}", retrieve(find_min(&tree)), retrieve(find_max(&tree)));
    } else {
        println!("Tree is empty");
    }

    tree = make_empty(tree);
}